### PR TITLE
Initial Structure

### PR DESCRIPTION
**Pull request by gblakeman**
A rough architecture for the plugin with some elements stubbed in. Some explanation and thoughts below.

#### manifest.json
Sets up our command/action listeners and maps them to `Core.js` to do the work. I’ve populated it with a small sample of commands and `OpenDocument` and `SelectionChange` action listeners. We get all of this for “free” from the Sketch API, but should be mindful of the resources we take up.

#### Core
A dispatcher of sorts; this centralizes the top-level scenarios we want the plugin to cover and sends most of the real work off to other classes.

I could eventually see this broken into several modules in different files depending on the features of the plugin. Lower-level logic would remain in the classes (below) and be easy to re-use.

#### Crawler
Will take in an array from a selection and help us iterate over it, returning useful items (layer, parent layer, artboard, document, etc.) depending on our needs.

#### Identifier
Will expect a single layer and based on the layer attributes (`id`, comparison with data source, etc.) help determine if the layer matches something in our design system and help keep track of it (i.e. match the layer with label layer(s) the plugin draws).

#### Painter ~~Drawing~~
Will handle any shape/object drawing we need to do _in_ the Sketch file to label design system components/foundations/etc.

#### Messenger
Will handle UI messages (toasts + alerts) and also logging. I could also see it handling any settings/prefs data unless that gets big enough to make sense standing on its own.

#### Tools
Re-usable bits that I expect we’ll create over time.

---

Related next steps: I don’t yet know enough about how we will get design system library data and use it to make our identifications, so I avoided making any assumptions here. For our MVP, it make be as simple as looking at data attached to a component when it’s imported into the Sketch file, or it may require something much more complex.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,6 +2,10 @@
   "name": "sketch-auto-spec",
   "description": "",
```
Comment (amiller-gh) [inline package.json]: Descriptions are nice :)

File: src/Drawing.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// will handle the tasks of drawing labels, and other adjustments
+// in the actual Sketch file
+import { Rectangle, ShapePath } from 'sketch/dom';
+
+export default class Drawing {
```
[the remark this answers is not in the document]
Comment (gblakeman) [reply src/Drawing.js]: Haha, yeah, I went back-and-forth on this one. I like `Painter`!

File: src/Core.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
+import { fromNative } from 'sketch';
+
+import Crawler from './Crawler';
+import Drawing from './Drawing';
+import Identifier from './Identifier';
+import Messenger from './Messenger';
+
+// import { setArray } from './Tools';
+
+const coreTasks = (() => {
```
[the remark this answers is not in the document]
Comment (gblakeman) [reply src/Core.js]: I wasn’t feeling that great about `coreTasks`. I like `mainElements` 👍👍

File: src/Core.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        //   messenger.log(firstSelectedItem);
+        // }
+      }
+      return null;
+    },
+  };
+})();
+
+// define each in manifest
+export const helloWorld = context => coreTasks.helloWorld(context);
```
[the remark this answers is not in the document]
Comment (gblakeman) [reply src/Core.js]: I ended up just simplifying this whole structure (now `main.js`) with a combined export.

File: src/Tools.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { toArray } from 'util';
+
+// conversion function to give us full js Array functions from an NSArray object
+export const setArray = nsArray => toArray(nsArray);
+
+export const hexToRgb = (hex) => {
```
[the remark this answers is not in the document]
Comment (gblakeman) [reply src/Tools.js]: Added documentation for everything. Let me know if I’m not adhering to any specific style guidelines here.

**Pull request merged**